lasagne/layers/recurrent.py

```diff
@@ -611,42 +602,59 @@ def step(input_dot_W_n, cell_previous, hid_previous):
 
-            cell, hid = step(input_dot_W_n, cell_previous, hid_previous)
+            cell, hid = step(input_dot_w_n, cell_previous, hid_previous)
 
             # If mask is 0, use previous state until mask = 1 is found.
             # This propagates the layer initial state when moving backwards
             # until the end of the sequence is found.
-            not_mask = 1 - mask
-            cell = cell*mask + cell_previous*not_mask
-            hid = hid*mask + hid_previous*not_mask
+            not_mask = 1 - mask_n
+            cell = cell*mask_n + cell_previous*not_mask
+            hid = hid*mask_n + hid_previous*not_mask
 
             return [cell, hid]
 
-        if self.backwards and mask is not None:
+        if mask is not None:
             # mask is given as (batch_size, seq_len). Because scan iterates
             # over first dimension, we dimshuffle to (seq_len, batch_size) and
             # add a broadcastable dimension
             mask = mask.dimshuffle(1, 0, 'x')
-            sequences = [input_dot_W, mask]
+            sequences = [input_dot_w, mask]
             step_fun = step_masked
         else:
-            sequences = input_dot_W
+            sequences = input_dot_w
             step_fun = step
 
         # Scan op iterates over first dimension of input and repeatedly
         # applies the step function
-        output = theano.scan(step_fun, sequences=sequences,
-                             outputs_info=[self.cell_init, self.hid_init],
-                             go_backwards=self.backwards,
-                             truncate_gradient=self.gradient_steps)[0][1]
-
-        # Now, dimshuffle back to (n_batch, n_time_steps, n_features))
-        output = output.dimshuffle(1, 0, 2)
-
-        # if scan is backward reverse the output
-        if self.backwards:
-            output = output[:, ::-1, :]
+        cell_out, hid_out = theano.scan(
+            step_fun,
+            sequences=sequences,
+            outputs_info=[self.cell_init, self.hid_init],
+            go_backwards=self.backwards,
+            truncate_gradient=self.gradient_steps)[0]
+
+        if self.return_sequence:
+            # dimshuffle back to (n_batch, n_time_steps, n_features))
+            hid_out = hid_out.dimshuffle(1, 0, 2)
+
+            # if scan is backward reverse the output
+            if self.backwards:
+                hid_out = hid_out[:, ::-1, :]
+        else:
+            # no need to dimshuffle because we return the last state
+            hid_out = hid_out[-1]
+        output = hid_out
+
+        # repeat for cell values if needed
+        if self.return_cell:
+            if self.return_sequence:
+                cell_out = cell_out.dimshuffle(1, 0, 2)
+                if self.backwards:
+                    cell_out = cell_out[:, ::-1, :]
+            else:
+                cell_out = cell_out[-1]
+            output = [cell_out, output]
```

Comment:
>> If we want to move the slicing outputside of the recurrent layers we can instead create a slicing layer.
>This sounds like the more general solution.

Agreed.  I think we should create a slicing layer and not have a `return_sequences` kwarg.

> I dont think that would solve the problem with the need to return both cell and hidden values.

I think it actually would (I may not have explained it well), but after considering it more I think it wouldn't allow for peephole connections, so it's not really a viable option.

> So maybe we should just add support for multiple-output layers, and the natural way to do so would be returning lists/tuples of outputs in get_output_for and get_output_shape_for. There could be other use cases for multiple-output layers. To allow stacking ordinary layers on top, we could have a DemuxLayer that just returns one of its multiple inputs. However, I think we shouldn't allow layers that either produce multiple outputs or not -- would they have a self.output_shape or a self.output_shapes property?

Agreed here in general.  I think in this particular case `self.output_shape` refers to the shape for both the cell and LSTM layer output, isn't that the case @skaae? But I could see that not being true for some other layer with multiple outputs.

@skaae, I'd like to merge your docstring audits/updates/tests/etc, but I don't think we've resolved discussion on the "output cell"/"layer with multiple outputs" question.  Can you move it to another PR so I can merge?

Thanks again.